File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/summary/SummaryViewModel.java
package ediger.diarynutrition.summary;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.Calendar;
import java.util.Collections;
import java.util.List;

import ediger.diarynutrition.AppContext;
import ediger.diarynutrition.data.DiaryRepository;
import ediger.diarynutrition.data.source.entities.Summary;
import ediger.diarynutrition.data.source.entities.Water;

public class SummaryViewModel extends AndroidViewModel {

    private int mNumOfDays = 1;

    private MutableLiveData<Calendar> mSinceDate = new MutableLiveData<>();

    private MediatorLiveData<List<Summary>> mSummary = new MediatorLiveData<>();

    private MediatorLiveData<List<Water>> mWater = new MediatorLiveData<>();

    private DiaryRepository mRepository;

    public SummaryViewModel(@NonNull Application application) {
        super(application);
        mRepository = ((AppContext) application).getRepository();

        mSummary.setValue(Collections.emptyList());
        mSummary.addSource(mSinceDate, date -> mSummary.setValue(mRepository.getSummary(date, mNumOfDays)));

        mWater.setValue(Collections.emptyList());
        mWater.addSource(mSinceDate, date -> mWater.setValue(mRepository.getWaterSummary(date, mNumOfDays)));
    }

    public LiveData<List<Summary>> getSummary() {
        return mSummary;
    }

    public LiveData<List<Water>> getWater() {
        return mWater;
    }

    public void setPeriod(Calendar since, int numberOfDays) {
        mNumOfDays = numberOfDays;
        mSinceDate.setValue(since);
    }
}

File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/summary/SummaryViewModel.kt
package ediger.diarynutrition.summary

import android.app.Application
import androidx.lifecycle.*
import ediger.diarynutrition.AppContext
import ediger.diarynutrition.data.source.entities.Summary
import ediger.diarynutrition.data.source.entities.Water
import kotlinx.coroutines.launch
import java.util.*

class SummaryViewModel(application: Application) : AndroidViewModel(application) {

    private val summaryRepository = (application as AppContext).summaryRepository

    private var numOfDays = 1

    private val sinceDate = MutableLiveData<Calendar>()

    private val _summary = MediatorLiveData<List<Summary>>()
    val summary: LiveData<List<Summary>> = _summary

    private val _water = MediatorLiveData<List<Water>>()
    val water: LiveData<List<Water>> = _water

    fun setPeriod(since: Calendar, numberOfDays: Int) {
        numOfDays = numberOfDays
        sinceDate.value = since
    }

    init {
        _summary.value = emptyList()
        _summary.addSource(sinceDate) { date: Calendar ->
            viewModelScope.launch {
                _summary.setValue(summaryRepository.getSummary(date, numOfDays))
            }
        }

        _water.value = emptyList()
        _water.addSource(sinceDate) { date: Calendar ->
            viewModelScope.launch {
                _water.setValue(summaryRepository.getWaterSummary(date, numOfDays))
            }
        }
    }
}